src/Agent.cs

```diff
@@ -0,0 +1,237 @@
+    /// <param name="value">The value to record.</param>
+    /// <param name="time">The exact time at which the event happened.  You probably want the default.</param>
+    /// <param name="count">The number of events which this represents.  You almost certainly want the default.</param>
+    /// <returns>The value being passed in via value, or null if something bad happened.</returns>
+    public float? Gauge(String metricName, float value, DateTime? time = null, int count = 1)
+    {
+      try
+        {
+          if(!ValidateMetricName(metricName)) return null;
+          if(!Enabled) return value;
+          int metricTime = (time ?? DateTime.Now).ToEpoch();
+          _collector.SendMessage(String.Format("gauge {0} {1} {2} {3}", metricName, value, metricTime, count));
+          return value;
+        }
+      catch (Exception e)
+        {
+          ReportException(e);
+        }
+      return null;
+    }
+
+    /// <summary>
+    ///   Measure the duration of an action in seconds.
+    /// </summary>
+    /// <param name="metricName">The name of the metric to send.  Letters, numbers, and dots.</param>
+    /// <param name="action">The action being measured</param>
+    /// <returns>The result of executing the action.</returns>
+    public T Time<T>(String metricName, Func<T> action)
+    {
+      return ActuallyTime(metricName, action);
+    }
+
+    /// <summary>
+    ///   Measure the duration of an action in seconds.
+    /// </summary>
+    /// <param name="metricName">The name of the metric to send.  Letters, numbers, and dots.</param>
+    /// <param name="action">The action being measured</param>
+    public void Time(String metricName, Action action)
+    {
+      ActuallyTime(metricName, action);
+    }
+
+    /// <summary>
+    ///   Measure the duration of an action in milliseconds.
+    /// </summary>
+    /// <param name="metricName">The name of the metric to send.  Letters, numbers, and dots.</param>
+    /// <param name="action">The action being measured</param>
+    /// <returns>The result of executing the action.</returns>
+    public T TimeMs<T>(String metricName, Func<T> action)
+    {
+      return ActuallyTime(metricName, action, 1000);
+    }
+
+    /// <summary>
+    ///   Measure the duration of an action in milliseconds.
+    /// </summary>
+    /// <param name="metricName">The name of the metric to send.  Letters, numbers, and dots.</param>
+    /// <param name="action">The action being measured</param>
+    public void TimeMs(String metricName, Action action)
+    {
+      ActuallyTime(metricName, action, 1000);
+    }
+
+    private T ActuallyTime<T>(String metricName, Func<T> action, float durationMultiplier = 1)
+    {
+      var start = DateTime.Now;
+      try
+        {
+          return action();
+        }
+      finally
+        {
+          var end = DateTime.Now;
+          var duration = end - start;
+          Gauge(metricName, (float)duration.TotalSeconds * durationMultiplier);
```

Comment:
I'm sure this is obvious if you csharp at all, but what's happening here? Is TotalSeconds something that's like a float but not a float? Is the casting necessary because of the multiplication?